app/models/conversation/block.rb

```diff
@@ -0,0 +1,5 @@
+class Conversation::Block < ApplicationRecord
```

Comment:
@kaspth, what's your rationale behind this over:

```ruby
module Conversations
  class Block

  # ...
```

I'm currently doing the `module` with a plural name so it doesn't conflict with the model name (e.g. `Businesses::Permission`) but really like this approach of not having to pluralize.

Is there a downside to doing this in one line?

Reply:
Whoh... This is really interesting. I didn't know that you could use the exact name of a class as a namespace if its a single line declaration. Is that a Zweitwerk related feature? What causes that difference in behavior?

I guess Ruby doesn't complain because we don't explicitly declare `Conversation` as a module. What assumption does it make about the kind of constant that `Conversation` is? If it assumes it's a class, why can't it cause a load sequence error like doing a multi-line declaration?

Reply:
> I hadn't thought about the fact that this wouldn't work if you're using the model name.

That's what I was afraid of and why I avoided the `Developer` _module` and went for `Developers` in the past. I really don't want to have a class be the namespacing because it requires an additional layer of complexity to understand what's going on.

But I do like that you can include the module without having to namespace it another layer!

Ideally the entire app follows the same convention (which is currently pluralized namespaces) but I do like how this feels a bit better. I'll have to think on it a bit more because the trade-offs might not be worth it.

Also super curious as to what @jacobdaddario is asking.

Reply:
> Is that a Zweitwerk related feature? What causes that difference in behavior?

It's a standard Ruby feature since is `Class` is a subclass of `Module`. The difference in behavior is due to loading order.

> I guess Ruby doesn't complain because we don't explicitly declare `Conversation` as a module. What assumption does it make about the kind of constant that `Conversation` is? If it assumes it's a class, why can't it cause a load sequence error like doing a multi-line declaration?

`Conversation` has already been loaded at this point, so Ruby doesn't assume anything. It's a class but classes can be namespaces too.

Reply:
> That's what I was afraid of and why I avoided the `Developer` module

Note: it wouldn't be a module though, it would still be the `Developer` class, but classes can be namespaces too.

> I really don't want to have a class be the namespacing because it requires an additional layer of complexity to understand what's going on.

What's the extra layer of complexity? To me it's more complex that I have to define a separate module — which exposes the namespace complexity anyhow, and could lead to loading order issues (though that doesn't happen with Zeitwerk).

> Ideally the entire app follows the same convention (which is currently pluralized namespaces) but I do like how this feels a bit better. I'll have to think on it a bit more because the trade-offs might not be worth it.

Looks like RuboCop has a rule that could be enforced for this (and the whole app could be converted in one-go if need be). https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Style/ClassAndModuleChildren

Reply:
Gotcha, so that makes more sense. It's inferring the one line assignment essentially as:
```
class Conversation
  class Block
  end
end
```
And the reason it has to be one-line is because of the load ordering. Sorry for the confusion around that. For some reason in my head the double semicolon always meant namespace, but obviously that's not true. It just means it's a constant holding another constant.